File: loadercomplex-fabric/src/ver122/java/com/enderzombi102/loadercomplex/fabric12/impl/world/FabricWorld.java
package com.enderzombi102.loadercomplex.fabric12.impl.world;

import com.enderzombi102.loadercomplex.api.minecraft.entity.PlayerEntity;
import com.enderzombi102.loadercomplex.api.minecraft.util.*;
import com.enderzombi102.loadercomplex.fabric12.impl.FabricServer;
import com.enderzombi102.loadercomplex.fabric12.impl.block.FabricBlockstate;
import com.enderzombi102.loadercomplex.api.minecraft.block.Blockstate;
import com.enderzombi102.loadercomplex.api.minecraft.entity.Entity;
import com.enderzombi102.loadercomplex.api.minecraft.world.World;
import com.enderzombi102.loadercomplex.fabric12.impl.utils.ConversionKt;
import net.minecraft.resource.Identifier;
import net.minecraft.util.math.BlockPos;

public class FabricWorld implements World {
	private final net.minecraft.world.World backingWorld;

	public FabricWorld( net.minecraft.world.World world ) {
		backingWorld = world;
	}

	@Override
	public void spawn( Entity entity, Position pos ) {
		entity.setPosition( pos );
		this.backingWorld.addEntity( (net.minecraft.entity.Entity) entity.getObject() );
	}

	@Override
	public boolean isClient() {
		return this.backingWorld.isClient;
	}

	@Override
	public boolean isHardcore() {
		return false;
	}

	@Override
	public boolean isSunny() {
		return this.backingWorld.isSunny();
	}

	@Override
	public boolean isRaining() {
		return this.backingWorld.isRaining();
	}

	@Override
	public boolean isThundering() {
		return this.backingWorld.isThundering();
	}

	@Override
	public boolean isAir( Position pos ) {
		return this.backingWorld.isAir( new BlockPos( pos.x, pos.y, pos.z ) );
	}

	@Override
	public boolean isPositionLoaded( Position pos ) {
		return this.backingWorld.isChunkLoaded( new BlockPos( pos.x, pos.y, pos.z ), true );
	}

	@Override
	public boolean hasBlockEntity( Position pos ) {
		return this.backingWorld.getBlockEntity( new BlockPos( pos.x, pos.y, pos.z ) ) != null;
	}

	@Override
	public boolean hasSkyAccess( Position pos ) {
		return this.backingWorld.hasSkyAccess( new BlockPos( pos.x, pos.y, pos.z ) );
	}

	@Override
	public boolean canSnowFall( Position pos ) {
		return this.backingWorld.canSnowFall( new BlockPos( pos.x, pos.y, pos.z ), true );
	}

	@Override
	public Blockstate getBlockState( Position pos ) {
		return new FabricBlockstate( this.backingWorld.getBlockState( new BlockPos( pos.x, pos.y, pos.z ) ) );
	}

	@Override
	public void setBlockState( Position pos, Blockstate state ) {
		this.backingWorld.setBlockState( new BlockPos( pos.x, pos.y, pos.z ), (net.minecraft.block.state.BlockState) state.getObject() );
	}

	@Override
	public void breakBlock( Position pos, boolean dropItems ) {
		this.backingWorld.breakBlock( new BlockPos( pos.x, pos.y, pos.z ), dropItems );
	}

	@Override
	public void removeBlock( Position pos ) {
		this.backingWorld.removeBlock( new BlockPos( pos.x, pos.y, pos.z ) );
	}

	@Override
	public Server getServer() {
		return new FabricServer( this.backingWorld.getServer() );
	}

	@Override
	public Position getSpawnLocation() {
		return ConversionKt.toLC( this.backingWorld.getSpawnPoint() );
	}

	@Override
	public Difficulty getDifficulty() {
		return Difficulty.valueOf( this.backingWorld.getDifficulty().name() );
	}

	@Override
	public int getRedstonePower( Position pos, Direction direction ) {
		return this.backingWorld.getDirectSignal(
			new BlockPos( pos.x, pos.y, pos.z ),
			net.minecraft.util.math.Direction.valueOf( direction.name() )
		);
	}

	@Override
	public void playsound( PlayerEntity player, double x, double y, double z, ResourceIdentifier sound, float volume, float pitch ) {
		net.minecraft.sound.SoundEvent event = net.minecraft.sound.SoundEvent.REGISTRY.get( new Identifier( sound.getNamespace(), sound.getPath() ) );
		if ( event == null ) {
			throw new IllegalArgumentException( String.format( "SoundEvent \"%s\" was not found!", sound ) );
		}
		this.backingWorld.playSound(
			(net.minecraft.entity.living.player.PlayerEntity) player.getObject(),
			x, y, z,
			event,
			net.minecraft.sound.SoundCategory.MASTER,
			volume, pitch
		);
	}

	@Override
	public Object getObject() {
		return this.backingWorld;
	}
}

File: loadercomplex-fabric/src/ver122/kotlin/com/enderzombi102/loadercomplex/fabric122/impl/world/FabricWorld.kt
package com.enderzombi102.loadercomplex.fabric122.impl.world

import com.enderzombi102.loadercomplex.api.math.Vec3d
import com.enderzombi102.loadercomplex.api.math.Vec3i
import com.enderzombi102.loadercomplex.api.minecraft.block.BlockState
import com.enderzombi102.loadercomplex.api.minecraft.entity.Entity
import com.enderzombi102.loadercomplex.api.minecraft.entity.PlayerEntity
import com.enderzombi102.loadercomplex.api.minecraft.util.Difficulty
import com.enderzombi102.loadercomplex.api.minecraft.util.ResourceIdentifier
import com.enderzombi102.loadercomplex.api.minecraft.server.Server
import com.enderzombi102.loadercomplex.fabric122.impl.server.FabricServer
import com.enderzombi102.loadercomplex.fabric122.impl.block.FabricBlockState
import com.enderzombi102.loadercomplex.fabric122.impl.utils.toLC
import com.enderzombi102.loadercomplex.fabric122.impl.utils.toMC
import net.minecraft.block.state.BlockState as McBlockState
import net.minecraft.resource.Identifier
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvent
import net.minecraft.world.World
import com.enderzombi102.loadercomplex.api.math.Direction as McDirection
import com.enderzombi102.loadercomplex.api.minecraft.world.World as McWorld
import net.minecraft.entity.Entity as McEntity


class FabricWorld( private val backingWorld: World ) : McWorld {
	override fun spawn( entity: Entity, pos: Vec3d ) {
		entity.position = pos
		backingWorld.addEntity( entity.getObject() as McEntity )
	}

	override fun isClient(): Boolean =
		backingWorld.isClient

	override fun isHardcore(): Boolean =
		false

	override fun isSunny(): Boolean =
		backingWorld.isSunny

	override fun isRaining(): Boolean =
		backingWorld.isRaining

	override fun isThundering(): Boolean =
		backingWorld.isThundering

	override fun isAir( pos: Vec3i ): Boolean =
		backingWorld.isAir( pos.toMC() )

	override fun isPositionLoaded( pos: Vec3i ): Boolean =
		backingWorld.isChunkLoaded( pos.toMC(), true )

	override fun hasBlockEntity( pos: Vec3i ): Boolean =
		backingWorld.getBlockEntity( pos.toMC() ) != null

	override fun hasSkyAccess( pos: Vec3i ): Boolean =
		backingWorld.hasSkyAccess( pos.toMC() )

	override fun canSnowFall( pos: Vec3i ): Boolean =
		backingWorld.canSnowFall( pos.toMC(), true )

	override fun getBlockState( pos: Vec3i ): BlockState =
		FabricBlockState( backingWorld.getBlockState( pos.toMC() ) )

	override fun setBlockState( pos: Vec3i, state: BlockState) =
		backingWorld.setBlockState( pos.toMC(), state.getObject() as McBlockState)

	override fun breakBlock( pos: Vec3i, dropItems: Boolean ) =
		backingWorld.breakBlock( pos.toMC(), dropItems )

	override fun removeBlock( pos: Vec3i ): Boolean =
		backingWorld.removeBlock( pos.toMC() )

	override fun getServer(): Server =
		FabricServer( backingWorld.server!! )

	override fun getSpawnLocation(): Vec3i =
		backingWorld.spawnPoint.toLC()

	override fun getDifficulty(): Difficulty =
		Difficulty.valueOf( backingWorld.difficulty.name )

	override fun getRedstonePower( pos: Vec3i, direction: McDirection ): Int {
		return backingWorld.getDirectSignal( pos.toMC(), direction.toMC() )
	}

	override fun playsound( player: PlayerEntity, x: Double, y: Double, z: Double, sound: ResourceIdentifier, volume: Float, pitch: Float ) {
		val event = SoundEvent.REGISTRY[Identifier( sound.namespace, sound.path )]
		requireNotNull( event ) { String.format( "SoundEvent \"%s\" was not found!", sound ) }
		backingWorld.playSound(
			player.toMC(),
			x, y, z,
			event,
			SoundCategory.MASTER,
			volume, pitch
		)
	}

	override fun getObject(): Any =
		this.backingWorld
}
